File: algorithm/cpp/closest-binary-search-tree-value.cpp
// 270. Closest Binary Search Tree Value
// Difficulty : Easy 

// Given a non-empty binary search tree and a target value, find the value in the BST that is closest to the target.

// Note:

// Given target value is a floating point.
// You are guaranteed to have only one unique value in the BST that is closest to the target.

// Time:  O(h)
// Space: O(1)

class Solution {
public:
  int closestValue(TreeNode* root, double target) {
    double gap = INT_MAX;
    int closest = INT_MAX;
    
    while (root) {
      if (abs(static_cast<double>(root->val) - target) < gap) {
        gap = abs(static_cast<double>(root->val) - target);
        closest = root->val;
      }
      if (target == root->val) {
        break;
      } else if (target < root->val) {
        root = root->left;
      } else {
        root = root->right;
      }
    }
    return closest;
  }
};

File: data-structure-and-algorithm/typescript/leetcode/src/closest-binary-search-tree-value.ts
/*
Given a non-empty binary search tree and a target value, find the value in the BST that is closest to the target.

Note:

Given target value is a floating point.
You are guaranteed to have only one unique value in the BST that is closest to the target.
Example:

Input: root = [4,2,5,1,3], target = 3.714286

    4
   / \
  2   5
 / \
1   3

Output: 4
*/

let result;
let min;

function closestValue(node, target) {
  min = Infinity;
  result = 0;
  dfs(node, target);
  return result;
}

function dfs(node, target) {
  if (!node) {
    return;
  }

  if (Math.abs(node.val - target) < min) {
    min = Math.abs(node.val - target);
    result = node.val;
  }

  if (target < node.val) {
    dfs(node.left, target);
  } else {
    dfs(node.right, target);
  }
}

function closestValueBST(root, target) {
  let result = root.val;
  while (root) {
    if (Math.abs(result - target) >= Math.abs(root.val - target)) {
      result = root.val;
    }
    root = target < root.val ? root.left : root.right;
  }
  return result;
}
